File: js/token-sdk/src/instructions/mint-to.ts
/**
 * Mint-to token instructions.
 */

import type { Address } from '@solana/addresses';
import {
    AccountRole,
    type Instruction,
    type AccountMeta,
} from '@solana/instructions';

import {
    DISCRIMINATOR,
    LIGHT_TOKEN_PROGRAM_ID,
    SYSTEM_PROGRAM_ID,
} from '../constants.js';
import { validatePositiveAmount, validateDecimals } from '../utils/validation.js';
import {
    getAmountInstructionEncoder,
    getCheckedInstructionEncoder,
    encodeMaxTopUp,
} from '../codecs/instructions.js';

/**
 * Parameters for minting tokens.
 */
export interface MintToParams {
    /** Mint address (CMint) */
    mint: Address;
    /** Token account to mint to */
    tokenAccount: Address;
    /** Mint authority - must be signer */
    mintAuthority: Address;
    /** Amount to mint */
    amount: bigint;
    /** Maximum lamports for rent top-up (optional, 0 = no limit) */
    maxTopUp?: number;
    /** Fee payer for rent top-ups (optional, defaults to authority) */
    feePayer?: Address;
}

/**
 * Creates a mint-to instruction (discriminator: 7).
 *
 * Mints tokens to a decompressed CToken account.
 *
 * Account layout:
 * 0: CMint account (writable)
 * 1: destination CToken account (writable)
 * 2: authority (signer, writable unless feePayer provided)
 * 3: system_program (readonly)
 * 4: fee_payer (optional, signer, writable)
 *
 * @param params - Mint-to parameters
 * @returns The mint-to instruction
 */
export function createMintToInstruction(params: MintToParams): Instruction {
    const { mint, tokenAccount, mintAuthority, amount, maxTopUp, feePayer } =
        params;

    validatePositiveAmount(amount);

    // Build accounts
    const accounts: AccountMeta[] = [
        { address: mint, role: AccountRole.WRITABLE },
        { address: tokenAccount, role: AccountRole.WRITABLE },
        {
            address: mintAuthority,
            role: feePayer
                ? AccountRole.READONLY_SIGNER
                : AccountRole.WRITABLE_SIGNER,
        },
        { address: SYSTEM_PROGRAM_ID, role: AccountRole.READONLY },
    ];

    // Add fee payer if provided
    if (feePayer) {
        accounts.push({ address: feePayer, role: AccountRole.WRITABLE_SIGNER });
    }

    // Build instruction data: discriminator + amount [+ maxTopUp]
    const baseBytes = getAmountInstructionEncoder().encode({
        discriminator: DISCRIMINATOR.MINT_TO,
        amount,
    });
    const maxTopUpBytes = encodeMaxTopUp(maxTopUp);

    const data = new Uint8Array(baseBytes.length + maxTopUpBytes.length);
    data.set(new Uint8Array(baseBytes), 0);
    if (maxTopUpBytes.length > 0) {
        data.set(maxTopUpBytes, baseBytes.length);
    }

    return {
        programAddress: LIGHT_TOKEN_PROGRAM_ID,
        accounts,
        data,
    };
}

/**
 * Parameters for mint-to checked.
 */
export interface MintToCheckedParams extends MintToParams {
    /** Expected decimals */
    decimals: number;
}

/**
 * Creates a mint-to checked instruction (discriminator: 14).
 *
 * Mints tokens with decimals validation.
 *
 * @param params - Mint-to checked parameters
 * @returns The mint-to checked instruction
 */
export function createMintToCheckedInstruction(
    params: MintToCheckedParams,
): Instruction {
    const {
        mint,
        tokenAccount,
        mintAuthority,
        amount,
        decimals,
        maxTopUp,
        feePayer,
    } = params;

    validatePositiveAmount(amount);
    validateDecimals(decimals);

    // Build accounts
    const accounts: AccountMeta[] = [
        { address: mint, role: AccountRole.WRITABLE },
        { address: tokenAccount, role: AccountRole.WRITABLE },
        {
            address: mintAuthority,
            role: feePayer
                ? AccountRole.READONLY_SIGNER
                : AccountRole.WRITABLE_SIGNER,
        },
        { address: SYSTEM_PROGRAM_ID, role: AccountRole.READONLY },
    ];

    // Add fee payer if provided
    if (feePayer) {
        accounts.push({ address: feePayer, role: AccountRole.WRITABLE_SIGNER });
    }

    // Build instruction data: discriminator + amount + decimals [+ maxTopUp]
    const baseBytes = getCheckedInstructionEncoder().encode({
        discriminator: DISCRIMINATOR.MINT_TO_CHECKED,
        amount,
        decimals,
    });
    const maxTopUpBytes = encodeMaxTopUp(maxTopUp);

    const data = new Uint8Array(baseBytes.length + maxTopUpBytes.length);
    data.set(new Uint8Array(baseBytes), 0);
    if (maxTopUpBytes.length > 0) {
        data.set(maxTopUpBytes, baseBytes.length);
    }

    return {
        programAddress: LIGHT_TOKEN_PROGRAM_ID,
        accounts,
        data,
    };
}

File: sdk-libs/token-pinocchio/src/instruction/mint_to.rs
//! MintTo CPI for Light Token operations.

use pinocchio::{
    account_info::AccountInfo,
    cpi::{slice_invoke, slice_invoke_signed},
    instruction::{AccountMeta, Instruction, Signer},
    program_error::ProgramError,
    pubkey::Pubkey,
};

use crate::constants::LIGHT_TOKEN_PROGRAM_ID;

/// Mint to ctoken via CPI.
///
/// # Example
///
/// ```rust,ignore
/// use light_token_pinocchio::instruction::MintToCpi;
///
/// MintToCpi {
///     mint: &ctx.accounts.mint,
///     destination: &ctx.accounts.destination,
///     amount: 100,
///     authority: &ctx.accounts.authority,
///     system_program: &ctx.accounts.system_program,
///     fee_payer: None,
/// }
/// .invoke()?;
/// ```
pub struct MintToCpi<'info> {
    pub mint: &'info AccountInfo,
    pub destination: &'info AccountInfo,
    pub amount: u64,
    pub authority: &'info AccountInfo,
    pub system_program: &'info AccountInfo,
    /// Optional fee payer for rent top-ups. If not provided, authority pays.
    pub fee_payer: Option<&'info AccountInfo>,
}

impl<'info> MintToCpi<'info> {
    pub fn invoke(self) -> Result<(), ProgramError> {
        self.invoke_signed(&[])
    }

    pub fn invoke_signed(self, signers: &[Signer]) -> Result<(), ProgramError> {
        // Build instruction data: discriminator(1) + amount(8) + optional max_top_up(2)
        let mut data = [0u8; 11];
        data[0] = 7u8; // MintTo discriminator
        data[1..9].copy_from_slice(&self.amount.to_le_bytes());
        let data_len = 9;

        // Authority is writable when no fee_payer is provided
        let authority_writable = self.fee_payer.is_none();

        let program_id = Pubkey::from(LIGHT_TOKEN_PROGRAM_ID);

        if let Some(fee_payer) = self.fee_payer {
            let account_metas = [
                AccountMeta::writable(self.mint.key()),
                AccountMeta::writable(self.destination.key()),
                if authority_writable {
                    AccountMeta::writable_signer(self.authority.key())
                } else {
                    AccountMeta::readonly_signer(self.authority.key())
                },
                AccountMeta::readonly(self.system_program.key()),
                AccountMeta::writable_signer(fee_payer.key()),
            ];

            let instruction = Instruction {
                program_id: &program_id,
                accounts: &account_metas,
                data: &data[..data_len],
            };

            let account_infos = [
                self.mint,
                self.destination,
                self.authority,
                self.system_program,
                fee_payer,
            ];

            if signers.is_empty() {
                slice_invoke(&instruction, &account_infos)
            } else {
                slice_invoke_signed(&instruction, &account_infos, signers)
            }
        } else {
            let account_metas = [
                AccountMeta::writable(self.mint.key()),
                AccountMeta::writable(self.destination.key()),
                if authority_writable {
                    AccountMeta::writable_signer(self.authority.key())
                } else {
                    AccountMeta::readonly_signer(self.authority.key())
                },
                AccountMeta::readonly(self.system_program.key()),
            ];

            let instruction = Instruction {
                program_id: &program_id,
                accounts: &account_metas,
                data: &data[..data_len],
            };

            let account_infos = [
                self.mint,
                self.destination,
                self.authority,
                self.system_program,
            ];

            if signers.is_empty() {
                slice_invoke(&instruction, &account_infos)
            } else {
                slice_invoke_signed(&instruction, &account_infos, signers)
            }
        }
    }
}
